Engine/BaseAssets/Components/SkeletalMeshComponent.cs

```diff
@@ -0,0 +1,295 @@
+                        {
+                            curIndex = i;
+                            break;
+                        }
+                    int nextIndex = curIndex + 1;
+
+                    float DeltaTime = curAnimation.RotationKeys[nextIndex].Time - curAnimation.RotationKeys[curIndex].Time;
+                    float Factor = (AnimationTime - curAnimation.RotationKeys[curIndex].Time) / DeltaTime;
+                    LinearAlgebra.Quaternion start = curAnimation.RotationKeys[curIndex].Rotation;
+                    LinearAlgebra.Quaternion end = curAnimation.RotationKeys[nextIndex].Rotation;
+                    rotation = LinearAlgebra.Quaternion.Slerp(start, end, Factor);
+                    rotation.normalize();
+                }
+
+                // position
+                Vector3f position;
+                if (curAnimation.PositionKeys.Count == 1)
+                    position = curAnimation.PositionKeys[0].Position;
+                else
+                {
+                    int curIndex = 0;
+                    for (int i = 0; i < curAnimation.PositionKeys.Count - 1; ++i)
+                        if (AnimationTime < curAnimation.PositionKeys[i + 1].Time)
+                        {
+                            curIndex = i;
+                            break;
+                        }
+                    int nextIndex = curIndex + 1;
+
+                    float DeltaTime = curAnimation.PositionKeys[nextIndex].Time - curAnimation.PositionKeys[curIndex].Time;
+                    float Factor = (AnimationTime - curAnimation.PositionKeys[curIndex].Time) / DeltaTime;
+                    Vector3f start = curAnimation.PositionKeys[curIndex].Position;
+                    Vector3f end = curAnimation.PositionKeys[nextIndex].Position;
+                    Vector3f delta = end - start;
+                    position = start + Factor * delta;
+                }
+
+                nodeTransform = Matrix4x4f.FromTranslation(position) * Matrix4x4f.FromQuaternion(rotation) * Matrix4x4f.FromScale(scaling);
+            }
+
+            Matrix4x4f globalTransform = parentTransform * nodeTransform;
+            BonesTransform[boneData.Index] = skeleton.InverseRootTransform * globalTransform * boneData.Offset;
+            InvTrspsBonesTransform[boneData.Index] = BonesTransform[boneData.Index].inverse().transposed();
+
+            foreach (int childIndex in boneData.ChildIndices)
+                UpdateTransform(skeleton, AnimationTime, skeleton.Bones[childIndex], globalTransform, BonesTransform, InvTrspsBonesTransform);
+        }
+
+        private void UpdateAnimation(out List<Matrix4x4f> BonesTransformations, out List<Matrix4x4f> InverseTransposeBonesTransformations)
+        {
+            BonesTransformations = new List<Matrix4x4f>();
+            InverseTransposeBonesTransformations = new List<Matrix4x4f>();
+            // calculate transform matrices
+            foreach (Skeleton skeleton in skeletons) {
```

Comment:
this foreach loop will fill BonesTransformations list with data from the last skeleton in the array, while processing all of them, better to use the first skeleton only (if any) and add todo to support multiple skeletons